File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.go
/*
func setZeroes(matrix [][]int)  {
    zeroRows := map[int]bool{}
    zeroCols := map[int]bool{}
    
    for i := 0; i < len(matrix); i++ {
        for j := 0; j < len(matrix[i]); j++ {
            if matrix[i][j] == 0 {
                zeroRows[i] = true
                zeroCols[j] = true
            }
        }
    }
    
    for i := 0; i < len(matrix); i++ {
        for j := 0; j < len(matrix[i]); j++ {
            if zeroRows[i] || zeroCols[j] {
                matrix[i][j] = 0
            }
        }
    }
    
    return
}
*/

func setZeroes(matrix [][]int)  {
    firstRowZero := false
    for _, v := range matrix[0] {
        if v == 0 {
            firstRowZero = true
            break
        }
    }
    
    for i := 1; i < len(matrix); i++ {
        for j := 0; j < len(matrix[i]); j++ {
            if matrix[i][j] == 0 {
                matrix[0][j] = 0
                matrix[i][0] = 0
            }
        }
    }
    
    for i := 1; i < len(matrix); i++ {
        if matrix[i][0] == 0 {
            for j := 0; j < len(matrix[i]); j++ {
                matrix[i][j] = 0
            }
        }
        for j := 0; j < len(matrix[i]); j++ {
            if matrix[0][j] == 0 {
                matrix[i][j] = 0
            }
        }
    }
    
    if firstRowZero {
        for i := 0; i < len(matrix[0]); i++ {
            matrix[0][i] = 0
        }
    }
    
    return
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
"""
# O(m + n)
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        zero_rows = set()
        zero_cols = set()
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                if matrix[i][j] == 0:
                    zero_rows.add(i)
                    zero_cols.add(j)
                    
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                if i in zero_rows or j in zero_cols:
                    matrix[i][j] = 0
        
        return
"""

class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        first_row_zero = 0 in matrix[0]
        
        for i in range(1, len(matrix)):
            for j in range(len(matrix[i])):
                if matrix[i][j] == 0:
                    matrix[0][j] = 0
                    matrix[i][0] = 0
        
        for i in range(1, len(matrix)):
            if 0 in matrix[i]:
                matrix[i] = [0] * len(matrix[i])
            else:    
                for j in range(0, len(matrix[i])):
                    print(f'(0, j): (0, {j})')
                    if matrix[0][j] == 0:
                        print(f'setting ({i}, {j}) to zero')
                        matrix[i][j] = 0
                        
        if first_row_zero:
            matrix[0] = [0] * len(matrix[0])
        
        return